lib/functoria.ml

```diff
@@ -21,39 +21,147 @@ open Functoria_misc
+  min : string option ;
+  max : string option
+}
+
+module Package = struct
+  (* we could have copied code from opam, but that's LGPL *)
+  let version_of_string s =
+    let r = List.fold_left (fun acc v ->
+        match Astring.String.to_int v with
+        | Some n -> n :: acc
+        | None -> invalid_arg "cannot parse version number")
+        []
+        (Astring.String.cuts ~sep:"." s)
+    in
+    List.rev r
+
+  let compare_version v1 v2 =
+    let rec cmp a b =
+      match a, b with
+      | x::xs, y::ys when x = y -> cmp xs ys
+      | x::_, y::_ -> compare x y
+      | [], y::ys when y = 0 -> cmp [] ys
+      | [], y::_ -> compare 0 y
+      | x::xs, [] when x = 0 -> cmp xs []
+      | x::_, [] -> compare x 0
+      | [], [] -> 0
+    in
+    let v1 = version_of_string v1
+    and v2 = version_of_string v2
+    in
+    cmp v1 v2
+
+  let m_option f a b = match a, b with
+    | None, None -> None
+    | Some a, None -> Some a
+    | None, Some b -> Some b
+    | Some a, Some b -> Some (f a b)
+
+  let merge opam a b =
+    let ocamlfind = String.Set.union a.ocamlfind b.ocamlfind
+    and min =
+      m_option
+        (fun a b -> match compare_version a b with 0 -> a | 1 -> a | _ -> b)
+        a.min b.min
+    and max =
+      m_option
+        (fun a b -> match compare_version a b with 0 -> a | 1 -> b | _ -> a)
+        a.max b.max
+    in
+    match min, max with
+    | Some a, Some b when compare_version a b >= 0 -> invalid_arg "version constraint must be that min is smaller than max"
+    | _ -> Some { opam ; build = a.build ; ocamlfind ; min ; max }
+
+  let package ?(build = false) ?sublibs ?ocamlfind ?min ?max opam =
+    let ocamlfind = match sublibs, ocamlfind with
+      | None, None -> [opam]
+      | Some xs, None -> opam :: List.map (fun x -> opam ^ "." ^ x) xs
+      | None, Some a -> a
+      | Some _, Some _ -> invalid_arg "only either ~sub or ~ocamlfind may be specified"
+    in
+    let ocamlfind = String.Set.of_list ocamlfind in
+    match min, max with
+    | Some min, Some max when compare_version min max >= 0 -> invalid_arg "min must be < max"
+    | _ -> { opam ; build ; ocamlfind ; min ; max }
+end
+
+let package = Package.package
 
+module Info = struct
   type t = {
     name: string;
     root: string;
     keys: Key.Set.t;
     context: Key.context;
-    libraries: String.Set.t;
-    packages: String.Set.t;
+    packages: package String.Map.t;
   }
 
   let name t = t.name
   let root t = t.root
-  let libraries t = String.Set.elements t.libraries
-  let packages t = String.Set.elements t.packages
+  let libraries t =
+    String.Set.elements
+      (List.fold_left String.Set.union String.Set.empty
+         (List.map (fun x -> x.ocamlfind)
+            (List.filter (fun x -> x.build = false)
+               (List.map snd (String.Map.bindings t.packages)))))
+  let package_names t =
+    List.map fst
+      (List.filter (fun (_, p) -> p.build = false)
+         (String.Map.bindings t.packages))
```

Comment:
Meh, `Package` is the module that deals with packages. You want something from `package list` to the list of ocamlfind (resp. opam packages), I feel like it would belong there.